### ISSUE TITLE
Better handle nullable values

### ISSUE THREAD

**Issue by playpauseandstop** [playpauseandstop/rororo#85]
Original report by @chobeat at https://github.com/playpauseandstop/rororo/issues/84#issuecomment-643498373,

> I think I've found an unrelated bug. In unmarshallers.py:198 it tries to access the keys of a None. From what I understand, it's doing recursion on nested dicts but it tries on the None value and it fails. This has been introduced between 2.0.0b3 and 2.0.0rc2. Before it was working properly.
> 
> I think it's happening on all my responses that contain at least a None value somewhere.

**Comment by playpauseandstop** [comment]
Hi @chobeat,

Let discuss the bug, unrelated to optimize `setup_openapi` here.

----

It is known problem with `openapi-core==0.13.3` that handling nullable values is broken. As all `rororo==2.0.0rcX` releases depend on `openapi-core==0.13.3` it means that you need to supply the request / response data knowing given limitation.

In my projects, to avoid this issue, I literally avoid using any nullables in request / response data. It is not pretty straightforward, but (from other angle) it allows me to reduce size of my request / responses.

However, I understand, that solution of ignoring nullables is not optimal. It requires to write on a regular basis cascade of `if` statements such as,

```py
async def some_view(request: web.Request) -> web.Response:
    data = await get_some_data()

    api_data = {
        "id": data.id, "title": data.title, "created_at": data.created_at.isoformat()
    }
    if data.description:
       api_data["description"] = data.description
    if data.modified_at:
       api_data["modified_at"] = data.modified_at.isoformat()

    return web.json_response(api_data)
```

With that in mind, I think `rororo` needs to not wait on new `openapi-core` release, but provide their own way of supporting nullable values. For doing that, I would really appreciate if you share with me your schema and data, so I can use it in `rororo` test cases.

----

If you interested, list of `openapi-core` issues about handling nullable data,

- Original issue, which is fixed in master, but not released at PyPI yet, https://github.com/p1c2u/openapi-core/issues/232
- New issue with nullable arrays, https://github.com/p1c2u/openapi-core/issues/251

**Comment by chobeat** [comment]
@playpauseandstop great. I will share with you a snippet the actual schema and payload. It's all part of public documentation of our product so I think there's no issue here. I'm sorry if I don't have time to come up with a minimal reproducible example.

```    SimulationList:
      properties:
        simulations:
          type: array
          items:
            $ref: '#/components/schemas/Simulation'```
```
     Simulation:
      properties:
        simulation_id:
          type: string
        encoded_payload_file_id:
          type: string
          nullable: true
        reconstructed_payload_file_id:
          type: string
          nullable: true
        dataset_id:
          type: string
        model_id:
          type: string
        simulation_metrics:
          type: object
          nullable: true
        batch_size:
          type: integer
          nullable: true```

Values:

```[{'dataset_id': 'test_dataset_id', 'batch_size': None, 'encoded_payload_file_id': None, 'model_id': 'test_model_id', 'reconstructed_dataset_file_id': None, 'simulation_id': '4c972a77-e86c-4d02-856f-dc1f3d2b9f5a', 'simulation_status': 'CREATED', 'simulation_metrics': None, 'simulation_errors': None}]```

Let me know if you manager to reproduce the error.  While extracting this information, I also noticed that the behavior is the same if the key is absent in the schema.

**Comment by playpauseandstop** [comment]
Thanks for sharing spec and example data with me! Will try to come up with solution to handle nullable data soon.

**Comment by playpauseandstop** [comment]
Hi @chobeat,

I've managed to fix issues with nullable arrays & objects, which not handled well by `openapi-core==0.13.3` in #87 

Shortly I plan to release `2.0.0rc3` version, _hopefully the last one before final `2.0.0` release_, so will appreciate if you will able to confirm whether my fixes help you or not.

**Comment by playpauseandstop** [comment]
https://pypi.org/project/rororo/2.0.0rc3/

**Comment by chobeat** [comment]
it works perfectly, both the new fix and the caching

**Comment by OysteinBrandt** [comment]
@playpauseandstop 
I have encountered another nullable value issue.
- rororo version: 2.0.2
- OAS version: 3.0.3
- when specifying `enum` and `nullable: true` in spec

#### minimal reproducible example
```yaml
# api.yaml
openapi: 3.0.3
info:
  version: 0.0.1
  title: example_of_nullable_enum
paths:
  '/configuration':
    get:
      operationId: get_config
      tags:
        - Configuration
      responses:
        '200':
          description: 'Successfully retrieved Configuration in json format'
          content:
            application/json:
              schema:
                type: object
                properties:
                  configuration1:
                    type: string
                    nullable: true # <-- the troublemaker
                    enum:
                      - val1
                      - val2
        default:
          description: 'Generic error'
          content:
            application/json:
              schema:
                type: object
                properties:
                  message:
                    type: string
```

```python
# main.py
from aiohttp import web
from rororo import (
    OperationTableDef,
    setup_openapi,
)
from rororo.openapi import get_validated_data, get_validated_parameters
from pathlib import Path

operations = OperationTableDef()

@operations.register
async def get_config(request: web.Request) -> web.Response:
  foo = { 'configuration1': None}
  return web.json_response(foo, status=200)

def main():
  app = setup_openapi(
    web.Application(),
    Path(__file__).parent / "api.yaml",
    operations,
    server_url="/",
  )

  web.run_app(app, port=5050)

if __name__ == '__main__':
  main()
```
A GET request on `http://localhost:5050/configuration` will then result in:
```json
{"detail": [{"loc": ["response", "configuration1"], "message": "None is not one of ['val1', 'val2']"}]}
```
